jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/actions/SyncServerlessAppAction.kt

```diff
@@ -0,0 +1,132 @@
+    null,
+    AwsIcons.Resources.SERVERLESS_APP
+) {
+    override fun actionPerformed(e: AnActionEvent) {
+        val project = e.getRequiredData(PlatformDataKeys.PROJECT)
+
+        if (!AwsConnectionManager.getInstance(project).isValidConnectionSettings()) {
+            notifyNoActiveCredentialsError(project = project)
+            return
+        }
+
+        ExecutableManager.getInstance().getExecutable<SamExecutable>().thenAccept { samExecutable ->
+            if (samExecutable is ExecutableInstance.InvalidExecutable || samExecutable is ExecutableInstance.UnresolvedExecutable) {
+                notifySamCliNotValidError(
+                    project = project,
+                    content = (samExecutable as ExecutableInstance.BadExecutable).validationError
+                )
+                return@thenAccept
+            }
+
+            val execVersion = SemVer.parseFromText(samExecutable.version) ?: error(" SAM CLI version could not detected")
+            val minVersion = SemVer("1.53.0", 1, 53, 0)
+            val minVersionForUseContainer = SemVer("1.57.0", 1, 57, 0)
+            if (!execVersion.isGreaterOrEqualThan(minVersion)) {
+                notifyError(message("sam.cli.version.warning"), message("sam.cli.version.upgrade.required"), project = project)
+                return@thenAccept
+            }
+
+            val templateFile = retrieveSamTemplate(e, project) ?: return@thenAccept
+
+            validateTemplateFile(project, templateFile)?.let {
+                notifyError(content = it, project = project)
+                return@thenAccept
+            }
+
+            val warningSettings = SamDisplayDevModeWarningSettings.getInstance()
+            runInEdt {
+                if (warningSettings.showDevModeWarning) {
+                    if (!SyncServerlessAppWarningDialog(project).showAndGet()) {
+                        return@runInEdt
+                    }
+                }
+
+                FileDocumentManager.getInstance().saveAllDocuments()
+                val parameterDialog = SyncServerlessApplicationDialog(project, templateFile)
+                if (!parameterDialog.showAndGet()) {
+                    // add telemetry
+                    return@runInEdt
+                }
+
+                val settings = parameterDialog.settings()
+
+                // TODO: saveSettings(project, templateFile, settings)
+
+                if (settings.useContainer) {
+                    if (!execVersion.isGreaterOrEqualThan(minVersionForUseContainer)) {
+                        notifyError(message("sam.cli.version.warning"), message("sam.cli.version.upgrade.required"), project = project)
+                        return@runInEdt
+                    }
+                    val dockerDoesntExist = runBlocking(getCoroutineBgContext()) {
+                        try {
+                            ExecUtil.execAndGetOutput(GeneralCommandLine("docker", "ps"))
+                            true
```

Comment:
do we need to check the exit code? what if they have docker installed but it's not running?

Reply:
Added an update to check exit codes along with exceptions